File: LeetCode/0347-Top K Frequent Elements/main.py
"""  
Given a non-empty array of integers, return the k most frequent elements.

Example 1:

Input: nums = [1,1,1,2,2,3], k = 2
Output: [1,2]
Example 2:

Input: nums = [1], k = 1
Output: [1]
Note:

You may assume k is always valid, 1 ≤ k ≤ number of unique elements.
Your algorithm's time complexity must be better than O(n log n), where n is the array's size.
It's guaranteed that the answer is unique, in other words the set of the top k frequent elements is unique.
You can return the answer in any order.
"""

from collections import Counter, defaultdict


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        topK = Counter(nums).most_common(k)  # [(num, freq)...]
        return [elem[0] for elem in topK]

    # 利用bucket sort
    """  
    freq_map保存 num -> 出现次数 的映射
    bucket是List[List], bucket的下标对应的是frequency
    bucket[i]保存了frequency=i的所有数字

    Time complexity: O(N)
    """

    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # index represents the frequency
        buckets = [[] for _ in range(len(nums) + 1)]
        num_freq = Counter(nums).items()

        for num, freq in num_freq:
            buckets[freq].append(num)

        res = []
        for item in buckets[::-1]:
            if item:
                for num in item:
                    res.append(num)
                    if len(res) == k:
                        return res

        return res

File: LeetCode/0347-Top K Frequent Elements/main.cpp
#include <iostream>
#include <map>
#include <queue>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> topKFrequent(vector<int>& nums, int k) {
    if (k == nums.size()) {
      return nums;
    }

    map<int, int> num_freq;
    for (int n : nums) {
      num_freq[n] += 1;
    }

    // custome compare function to initialzie min_heap
    auto cmp = [&num_freq](int n1, int n2) {
      return num_freq[n1] > num_freq[n2];
    };

    priority_queue<int, vector<int>, decltype(cmp)> pq(cmp);

    for (pair<int, int> p : num_freq) {
      pq.push(p.first);
      if (pq.size() > k) {
        pq.pop();  // pop the least frequency element
      }
    }

    vector<int> res;
    for (int i = 0; i < k; ++i) {
      res.push_back(pq.top());
      pq.pop();
    }

    return res;
  }

  vector<int> topKFrequent2(vector<int>& nums, int k) {
    if (k == nums.size()) {
      return nums;
    }

    map<int, int> num_freq;
    for (int n : nums) {
      num_freq[n] += 1;
    }

    // row: frequency; col: nums with the same freq
    vector<vector<int>> buckets((int)(nums.size() + 1));
    for (pair<int, int> p : num_freq) {
      buckets[p.second].push_back(p.first);
    }

    vector<int> res;
    for (auto it = buckets.rbegin(); it != buckets.rend(); ++it) {
      if (!(*it).empty()) {
        for (int num : (*it)) {
          res.push_back(num);
          if (--k == 0) {
            return res;
          }
        }
      }
    }

    return res;
  }
};
